Develop

oaftobar opened the pull request:
Initial code import

oaftobar committed:
Add project files to develop branch

oaftobar committed:
Update inventory.txt

JohnZavisa reviewed (commented):
Please pull down this repo and make sure it runs properly @oaftobar

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+#   -[x] `list_low_stock_products(self, threshold)`: Returns a list of products with quantity below the given threshold
+
+class InventoryManager:
+    def __init__(self, filename="InventoryManager/inventory.txt"):
```
This is a bug

JohnZavisa commented on .DS_Store:
Please do not commit this file to the repo, its for your local machine only

oaftobar committed:
Delete .DS_Store

Local machine file accidentally uploaded.

oaftobar committed:
Update InventoryManagementSystem.py

Removed "inventory manager" prefix from file path as inventory.txt is in the same directory.

JohnZavisa commented:
`
# Current implementation has everything in one file
# We should separate into:
├── inventory_management/
│   ├── __init__.py
│   ├── models/
│   │   ├── __init__.py
│   │   ├── product.py
│   │   └── category.py
│   ├── managers/
│   │   ├── __init__.py
│   │   └── inventory_manager.py
│   └── cli/
│       └── main.py
└── tests/
    └── test_inventory_manager.py
    `

JohnZavisa commented:
```python
# Current implementation has everything in one file
# We should separate into:
├── inventory_management/
│   ├── __init__.py
│   ├── models/
│   │   ├── __init__.py
│   │   ├── product.py
│   │   └── category.py
│   ├── managers/
│   │   ├── __init__.py
│   │   └── inventory_manager.py
│   └── cli/
│       └── main.py
└── tests/
    └── test_inventory_manager.py
```

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+#   -[x] View the total inventory value
+#   -[x] List low stock products
+
+manager = InventoryManager()
```
```python
# Current: Code runs at module level
manager = InventoryManager()
while True:
    # ... menu code ...

# Should be:
def main():
    manager = InventoryManager()
    # ... menu code ...

if __name__ == "__main__":
    main()
```

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+    def __init__(self, filename="inventory.txt"):
+        self.filename = filename
+        self.Inventory = self.load_inventory()
+        #self.Inventory = [] # Empty list to hold Inventory
```
```python
# Current:
self.Inventory  # Incorrect: Capital letter in instance variable

# Should be:
self.inventory  # Correct: snake_case for variables
```

JohnZavisa replied:
Please remove commented out code as well

oaftobar replied:
I'm guessing I was supposed to change all instances of "Inventory" to "inventory" as when I did those pieces started working

oaftobar replied:
Should I remove ALL commented out code like the notes I had for tasks??

JohnZavisa replied:
No, just unused code, you can keep notes for now until we're closer to being done

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+            print(f"{self.filename} not found.")
+        return inventory
+    
+    def save_inventory(self):
```
This is unsafe because:

If the program crashes during writing (power failure, system error, etc.)
If disk space runs out
If permissions change mid-write
If another process tries to read the file while it's being written

You could end up with:

A partially written file
Corrupted data
Complete data loss
Inconsistent state

Here's how to make it atomic (all-or-nothing):

```python
# Current
def save_inventory(self):
    with open(self.filename, "w") as file:
        for product in self.Inventory:
            file.write(f'{product}\n')

# Recommended
from pathlib import Path
import tempfile
import shutil

def save_inventory(self):
    """Save inventory to file with atomic write operation."""
    inventory_path = Path(self.filename)
    # Create directory if it doesn't exist
    inventory_path.parent.mkdir(parents=True, exist_ok=True)
    
    # Use temporary file for atomic write
    with tempfile.NamedTemporaryFile(mode='w', delete=False) as tmp_file:
        try:
            for product in self.Inventory:
                tmp_file.write(f'{product}\n')
            tmp_file.flush()
            shutil.move(tmp_file.name, self.filename)
        except Exception as e:
            Path(tmp_file.name).unlink(missing_ok=True)
            raise IOError(f"Failed to save inventory: {e}")
``` 

This approach is safer because:

Writes to a temporary file first
Only moves the file when writing is complete (move operations are atomic on most systems)
Keeps a backup of the previous version
Cleans up on failure
Restores from backup if something goes wrong

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+    def get_total_inventory_value(self):
+        return sum(product.get_value() for product in self.Inventory)
+
+    def list_low_stock_products(self, threshold=5):
```
Add type hints throughout, example:

```python
from typing import List, Optional

class InventoryManager:
    def list_low_stock_products(self, threshold: int = 5) -> List[str]:
        return [product.name for product in self.inventory if product.quantity < threshold]
```

oaftobar replied:
This has been completed with 5th iteration

JohnZavisa commented:
> ```python
> # Current implementation has everything in one file
> # You should separate into:
> ├── inventory_management/
> │   ├── __init__.py
> │   ├── models/
> │   │   ├── __init__.py
> │   │   ├── product.py
> │   │   └── category.py
> │   ├── managers/
> │   │   ├── __init__.py
> │   │   └── inventory_manager.py
> │   └── cli/
> │       └── main.py
> └── tests/
>     └── test_inventory_manager.py
> ```

Example test file

```python
# test_inventory_manager.py
import pytest
from pathlib import Path
from inventory_management.models.product import Product
from inventory_management.managers.inventory_manager import InventoryManager

@pytest.fixture
def temp_inventory_file(tmp_path):
    return tmp_path / "test_inventory.txt"

@pytest.fixture
def manager(temp_inventory_file):
    return InventoryManager(temp_inventory_file)

def test_add_product(manager):
    product = Product("Test Item", 10.00, 5)
    manager.add_product(product)
    assert len(manager.inventory) == 1
    assert manager.inventory[0] == product
```

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+    def __repr__(self):
+        return f'{self.name}, {self.price}, {self.quantity}'
+    
+    @staticmethod
```
What is this annotation for?

oaftobar replied:
Do you mean the **staticmethod**? 
This is so I don't have to have the parsing logic within _load_inventory_ and is encapsulated within the Product class so if I need to modify anything it can be done from within there as well, as well as being able to easily reuse the _from_string_ elsewhere.

JohnZavisa replied:
Here is a bit of a better answer, make sure you understand fundamentally that the point of a static method is that it doesn't need to access attributes of the class object.

"I used @staticmethod because converting a string into a Product is an operation that doesn't need access to any instance data (doesn't need self). It's a standalone conversion utility that logically belongs to the Product class but doesn't need to interact with any existing Product object's attributes. This is different from methods like get_value() which need access to the instance's price and quantity through self."

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+        self.Inventory.append(product)
+        self.save_inventory()
+        print(f"Product: {product.name} (x{product.quantity}) has been added at ${product.price}")
+        return
```
Remove unnecessary returns, this function does not return anything

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+            for product in self.Inventory:
+                file.write(f'{product}\n')
+    
+    def add_product(self, product):
```
add proper logging please, example:

```python
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class InventoryManager:
    def add_product(self, product):
        logger.info(f"Adding product: {product.name}")
        # ... rest of the method
```

oaftobar replied:
Does the **import** and following _logging_ and _logger_ sections go towards the very top of the file?

JohnZavisa replied:
Yes

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+#   -[x] `get_total_inventory_value(self)`: Calculates and returns the total value of all products in the inventory
+#   -[x] `list_low_stock_products(self, threshold)`: Returns a list of products with quantity below the given threshold
+
+class InventoryManager:
```
Add doc strings please:

```python
class InventoryManager:
    """Manages inventory of products with file persistence.
    
    Attributes:
        filename (str): Path to inventory file
        inventory (List[Product]): List of products in inventory
    """
```

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,198 @@
+                return
+        print(f"Product {product_name} not found in inventory.")
+
+    def update_quantity(self, product_name, new_quantity):
```
add method strings please:

```python
def update_quantity(self, product_name: str, new_quantity: int) -> None:
    """Updates the quantity of a product in inventory.
    
    Args:
        product_name: Name of product to update
        new_quantity: New quantity to set
        
    Raises:
        ValueError: If new_quantity is negative
    """
```

JohnZavisa replied:
Please add to each method

oaftobar replied:
Docstrings have been added to all methods

oaftobar committed:
[feature] - 1st iteration addressing changes

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,237 @@
+                for line in file:
+                    inventory.append(Product.from_string(line))
+        except FileNotFoundError:
+            print(f"{self.filename} not found.")
```
here you would want to use the error logger, like the info thing you already did. Remove the Print

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,237 @@
+            if product.name == product_name:
+                self.inventory.remove(product)
+                self.save_inventory()
+                print(f"Removed {product_name}")
```
use the logger instead of print

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,237 @@
+                self.save_inventory()
+                print(f"Removed {product_name}")
+                return
+        print(f"Product {product_name} not found in inventory.")
```
logger not print

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,237 @@
+            if product.name == product_name:
+                product.quantity = new_quantity
+                self.save_inventory()
+                print(f"Product {product.name} has been updated to new quantity:  {product.quantity} ")
```
logger not print

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,237 @@
+                self.save_inventory()
+                print(f"Product {product.name} has been updated to new quantity:  {product.quantity} ")
+                return
+        print(f"Product {product_name} not found.")
```
same as above

oaftobar committed:
[feature] - 2nd iteration addressing changes

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -72,7 +70,7 @@ def load_inventory(self):
                     inventory.append(Product.from_string(line))
         except FileNotFoundError:
-            print(f"{self.filename} not found.")
+            logger.info(f'{self.filename} not found.')
```
lets use logger.error for this one

oaftobar committed:
[feature] - 3rd iteration addressing changes

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,235 @@
+                print(f"{product.name} already exists. Please enter a new product, or select '3' to update quantity.")
+        self.inventory.append(product)
+        self.save_inventory()
+        print(f"Product: {product.name} (x{product.quantity}) has been added at ${product.price}")
```
Logger not print

oaftobar replied:
Changed print instance to logger.info

JohnZavisa commented on InventoryManagementSystem.py:
```diff
@@ -0,0 +1,235 @@
+        for existing_product in self.inventory:
+            logger.info(f'Adding product: {product.name}')
+            if product.name == existing_product.name:
+                print(f"{product.name} already exists. Please enter a new product, or select '3' to update quantity.")
```
logger not print

oaftobar replied:
Changed print instance to logger.info

oaftobar committed:
[feature] - 4th iteration addressing changes

oaftobar commented:
### 

> > ```python
> > # Current implementation has everything in one file
> > # You should separate into:
> > ├── inventory_management/
> > │   ├── __init__.py
> > │   ├── models/
> > │   │   ├── __init__.py
> > │   │   ├── product.py
> > │   │   └── category.py
> > │   ├── managers/
> > │   │   ├── __init__.py
> > │   │   └── inventory_manager.py
> > │   └── cli/
> > │       └── main.py
> > └── tests/
> >     └── test_inventory_manager.py
> > ```
> 
> Example test file
> 
> ```python
> # test_inventory_manager.py
> import pytest
> from pathlib import Path
> from inventory_management.models.product import Product
> from inventory_management.managers.inventory_manager import InventoryManager
> 
> @pytest.fixture
> def temp_inventory_file(tmp_path):
>     return tmp_path / "test_inventory.txt"
> 
> @pytest.fixture
> def manager(temp_inventory_file):
>     return InventoryManager(temp_inventory_file)
> 
> def test_add_product(manager):
>     product = Product("Test Item", 10.00, 5)
>     manager.add_product(product)
>     assert len(manager.inventory) == 1
>     assert manager.inventory[0] == product
> ```

Let me know when you're free to go over this 1v1

oaftobar committed:
[feature] 5th iteration - added docstrings everywhere possible, separated out original code to folder & subfolders.

oaftobar committed:
[feature] - 6th iteration, correct separation & dynamically wrote inventory.txt, removing caches

oaftobar committed:
Remove __pycache__ files, adding .gitignore
